### PR TITLE
Enhance media protections

### LINKED ISSUE #481: BUG: `MessageIsMedia` does not check `m.new_content`
When blocking messages there is media, it is possible to bypass if editing the message.
This can be done by sending a new `m.room.message` with `m.new_content` being a image type instead (`m.image`).
This will get posted, without mjolnir reacting to it, and blocking it :)

It seems like `MessageIsMedia` should recurse into messages, to check for such things.

Hopefully this is the correct way to report it!

EDIT: should clarify, this bypasses everything in the current checker, unless the checker recurses into the message.

### PR DESCRIPTION
**Pull request by H-Shay**
As the title states.

Fixes #481

### CODE REVIEW

File: src/protections/MessageIsMedia.ts [hunk 1]
```diff
@@ -35,10 +35,11 @@ export class MessageIsMedia extends Protection {
 
     public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any): Promise<any> {
         if (event['type'] === 'm.room.message') {
-            const content = event['content'] || {};
+            let content = event['content'] || {};
+            content = content?.["m.new_content"] ?? content;
```
Comment (turt2live) [inline src/protections/MessageIsMedia.ts:39]: we should probably also check `m.relates_to` to ensure we're actually looking at an edit here. The risk being that someone decides to send partially malformed events to trick Mjolnir into (dis)allowing something which is still rendered to users.

File: src/protections/MessageIsMedia.ts [hunk 2]
```diff
@@ -35,10 +35,11 @@ export class MessageIsMedia extends Protection {
...
             const msgtype = content['msgtype'] || 'm.text';
             const formattedBody = content['formatted_body'] || '';
-            const isMedia = msgtype === 'm.image' || msgtype === 'm.video' || msgtype === 'm.sticker' || formattedBody.toLowerCase().includes('<img');
+            let isMedia = msgtype === 'm.image' || msgtype === 'm.video' || msgtype === 'm.sticker' || formattedBody.toLowerCase().includes('<img');
```
Comment (turt2live) [inline src/protections/MessageIsMedia.ts:42]: this doesn't appear to need a const->let change

**Review by turt2live — CHANGES_REQUESTED** [review]
otherwise lgtm - thanks!

File: src/protections/MessageIsMedia.ts (hunk 1 above)
Comment (H-Shay) [reply src/protections/MessageIsMedia.ts:39]: So we can assume it's an edit if there is both an `m.new_content` and an `m.relates_to` field?

Comment (turt2live) [reply src/protections/MessageIsMedia.ts:39]: `m.relates_to` can mean [many things](https://spec.matrix.org/v1.11/client-server-api/#relationship-types), but inspecting it should tell us if it's an edit. In practice this means ensuring the `m.relates_to` `rel_type` is `m.replace`, I think. There's a bunch of [validity rules](https://spec.matrix.org/v1.11/client-server-api/#validity-of-replacement-events) we should probably be checking too, but that might still have false negatives if the code isn't perfect (or someone finds a way to send something weird).

File: src/protections/MessageIsMedia.ts [hunk 3]
```diff
@@ -35,7 +35,11 @@ export class MessageIsMedia extends Protection {
 
     public async handleEvent(mjolnir: Mjolnir, roomId: string, event: any): Promise<any> {
         if (event['type'] === 'm.room.message') {
-            const content = event['content'] || {};
+            let content = event['content'] || {};
+            const relation = content["m.relates_to"]
+            if (relation && relation["rel_type"] === "m.replace") {
```
Comment (turt2live) [inline src/protections/MessageIsMedia.ts:40]: fwiw, this should resolve to the same thing:

```suggestion
            if (relation?.["rel_type"] === "m.replace") {
```

**Review by turt2live — APPROVED** [review]
lgtm - thanks!

**Pull request merged**